python/pyspark/pandas/groupby.py

```diff
@@ -2732,34 +2741,33 @@ def stat_function(col: Column) -> Column:
             return F.percentile_approx(col, 0.5, accuracy)
 
         return self._reduce_for_stat_function(
-            stat_function, only_numeric=numeric_only, bool_to_numeric=True
+            stat_function,
+            accepted_spark_types=(NumericType, BooleanType) if numeric_only else None,
+            bool_to_numeric=True,
         )
 
     def _reduce_for_stat_function(
         self,
         sfun: Callable[[Column], Column],
-        only_numeric: Optional[bool] = None,
-        bool_as_numeric: bool = False,
+        accepted_spark_types: Optional[Tuple[Type[DataType], ...]] = None,
         bool_to_numeric: bool = False,
     ) -> FrameLike:
         """Apply an aggregate function `sfun` per column and reduce to a FrameLike.
 
         Parameters
         ----------
         sfun : The aggregate function to apply per column
-        only_numeric: If True, only numeric columns are involved
-        bool_as_numeric: If True, boolean columns are seen as numeric columns (following pandas)
+        accepted_spark_types: Accepted spark types of columns to be aggregated;
+                              default None means all spark types are accepted
         bool_to_numeric: If True, boolean columns are converted to numeric columns
         """
         groupkey_names = [SPARK_INDEX_NAME_FORMAT(i) for i in range(len(self._groupkeys))]
         groupkey_scols = [s.alias(name) for s, name in zip(self._groupkeys_scols, groupkey_names)]
 
         agg_columns = []
         for psser in self._agg_columns:
-            if (
-                isinstance(psser.spark.data_type, NumericType)
-                or (bool_as_numeric and isinstance(psser.spark.data_type, BooleanType))
-                or not only_numeric
+            if (accepted_spark_types is None) or isinstance(
+                psser.spark.data_type, accepted_spark_types
             ):
                 agg_columns.append(psser)
             elif bool_to_numeric and isinstance(psser.spark.data_type, BooleanType):
```

Comment:
Is this if-else order correct?
What if `accepted_spark_types` includes `BooleanType` and `bool_to_numeric` is `True`? e.g., https://github.com/apache/spark/blob/6a0e11a7d93ef0d583c226cab72bb96f8641551a/python/pyspark/pandas/groupby.py#L2743-L2747

Reply:
Good catch! That made me realize there is an overlap between semantics of `accepted_spark_types` and `bool_to_numeric`. The docstring is adjusted for that as well.

Reply:
We should also fix the above example to be clearer?
Now we don't need to specify `BooleanType` in `accepted_spark_types`, right?

Reply:
Sounds good! Modified.